local inventory = require("./inventory")
local catalog = require("./catalog")
local avatar = require("./avatar")
local premiumfeatures = require("./premiumfeatures")
local presence = require("./presence")
local chat = require("./chat")
local trades = require("./trades")
local accountsettings = require("./accountsettings")
local groups = require("./groups")
local economy = require("./economy")
local privatemessages = require("./privatemessages")
local develop = require("./develop")
local friends = require("./friends")
local users = require("./users")
local badges = require("./badges")
local notifications = require("./notifications")
local games = require("./games")
local thumbnails = require("./thumbnails")
local followings = require("./followings")

return {
	inventory = inventory,
	catalog = catalog,
	avatar = avatar,
	premiumfeatures = premiumfeatures,
	presence = presence,
	chat = chat,
	trades = trades,
	accountsettings = accountsettings,
	groups = groups,
	economy = economy,
	privatemessages = privatemessages,
	develop = develop,
	friends = friends,
	users = users,
	badges = badges,
	notifications = notifications,
	games = games,
	thumbnails = thumbnails,
	followings = followings,
}
